### ISSUE TITLE
How to build on Windows10?

### ISSUE THREAD

**Issue by WhiteEyeYan** [sowson/darknet#42]
I success used Cmake to make a .sln file. 
And I'm trying to compile it by Visual Studio 2019.
But it has too much error, after I fix one more error come out.
Could anyone help me build it on windwos?

**Comment by sowson** [comment]
Please make a fork and try hard... I am really interested in integrating such code to allow the Win10 build to succeed. Thanks!

**Comment by WhiteEyeYan** [comment]
Sorry, I'm not good at programming but I'll keep trying.
Now, I have another problem.
I build on ubuntu 18.04 with AMD RX480.
I set yolov4-tiny.cfg like this:
> [net]
>#Testing
> #batch=1
> #subdivisions=1
> #Training
> batch=1
> subdivisions=1
> width=32
> height=32
> channels=3
> momentum=0.9
> decay=0.0005
> angle=0
> saturation = 1.5
> exposure = 1.5
> hue=.1
> flip=0
> learning_rate=0.00261
> burn_in=1000
> max_batches = 8000
> policy=steps
> steps=6400,7200
> scales=.1,.1
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=3
> stride=2
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=2
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers=-1
> groups=2
> group_id=1
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -1,-2
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -6,-1
> 
> [maxpool]
> size=2
> stride=2
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers=-1
> groups=2
> group_id=1
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -1,-2
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -6,-1
> 
> [maxpool]
> size=2
> stride=2
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers=-1
> groups=2
> group_id=1
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -1,-2
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -6,-1
> 
> [maxpool]
> size=2
> stride=2
> 
> [convolutional]
> batch_normalize=1
> filters=512
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> ##################################
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=512
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> size=1
> stride=1
> pad=1
> filters=27
> activation=linear
> 
> 
> 
> [yolo4]
> mask = 3,4,5
> anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
> classes=4
> num=6
> jitter=.3
> scale_x_y = 1.05
> cls_normalizer=1.0
> iou_normalizer=0.07
> iou_loss=ciou
> ignore_thresh = .7
> truth_thresh = 1
> random=0
> resize=1.5
> nms_kind=greedynms
> beta_nms=0.6
> 
> [route]
> layers = -4
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [upsample]
> stride=2
> 
> [route]
> layers = -1, 23
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> size=1
> stride=1
> pad=1
> filters=27
> activation=linear
> 
> [yolo4]
> mask = 1,2,3
> anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
> classes=4
> num=6
> jitter=.3
> scale_x_y = 1.05
> cls_normalizer=1.0
> iou_normalizer=0.07
> iou_loss=ciou
> ignore_thresh = .7
> truth_thresh = 1
> random=0
> resize=1.5
> nms_kind=greedynms
> beta_nms=0.6

It will have error "Segmentation fault (core dumped)".
![2021-02-02 14-18-48 的螢幕擷圖](https://user-images.githubusercontent.com/67775380/106560417-f6b19b80-6561-11eb-809b-2b4fa8d5c085.png)

If I set subdivisions > batch like this:
> [net]
>#Testing
> #batch=1
> #subdivisions=1
> #Training
> batch=1
> subdivisions=2
> width=32
> height=32
> channels=3
> momentum=0.9
> decay=0.0005
> angle=0
> saturation = 1.5
> exposure = 1.5
> hue=.1
> flip=0
> learning_rate=0.00261
> burn_in=1000
> max_batches = 8000
> policy=steps
> steps=6400,7200
> scales=.1,.1
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=3
> stride=2
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=2
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers=-1
> groups=2
> group_id=1
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -1,-2
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -6,-1
> 
> [maxpool]
> size=2
> stride=2
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers=-1
> groups=2
> group_id=1
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -1,-2
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -6,-1
> 
> [maxpool]
> size=2
> stride=2
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers=-1
> groups=2
> group_id=1
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -1,-2
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [route]
> layers = -6,-1
> 
> [maxpool]
> size=2
> stride=2
> 
> [convolutional]
> batch_normalize=1
> filters=512
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> ##################################
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=512
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> size=1
> stride=1
> pad=1
> filters=27
> activation=linear
> 
> 
> 
> [yolo4]
> mask = 3,4,5
> anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
> classes=4
> num=6
> jitter=.3
> scale_x_y = 1.05
> cls_normalizer=1.0
> iou_normalizer=0.07
> iou_loss=ciou
> ignore_thresh = .7
> truth_thresh = 1
> random=0
> resize=1.5
> nms_kind=greedynms
> beta_nms=0.6
> 
> [route]
> layers = -4
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [upsample]
> stride=2
> 
> [route]
> layers = -1, 23
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> size=1
> stride=1
> pad=1
> filters=27
> activation=linear
> 
> [yolo4]
> mask = 1,2,3
> anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
> classes=4
> num=6
> jitter=.3
> scale_x_y = 1.05
> cls_normalizer=1.0
> iou_normalizer=0.07
> iou_loss=ciou
> ignore_thresh = .7
> truth_thresh = 1
> random=0
> resize=1.5
> nms_kind=greedynms
> beta_nms=0.6

error become
 "FATAL ERROR: CL_INVALID_BUFFER_SIZE
could create buffer on device. error: CL_INVALID_BUFFER_SIZE"
![2021-02-02 14-27-01 的螢幕擷圖](https://user-images.githubusercontent.com/67775380/106560895-c28aaa80-6562-11eb-8924-9c7d0cf4c097.png)

I have check my label.txt without 0.0, every number is larger than 0.
I also try
> export CPU_MAX_ALLOC_PERCENT=100
> export GPU_SINGLE_ALLOC_PERCENT=100
> export GPU_MAX_ALLOC_PERCENT=100
> export GPU_MAX_HEAP_SIZE=100

Do you have any suggestions to slove this?

**Comment by achateigner** [comment]
Hi, I am interested in building it for windows too.
Are there any libs that you are using that are specific to unix systems?

**Comment by elad8a** [comment]
@sowson I'm interested too, currently trying to resolve the cmake issues. it might be beneficial to try and write up a simple build instructions that lists dependencies.

currently, I struggle to configure clblas. but I also see the usage of non portable includes in the src files. pthreads is one example.

**Comment by sowson** [comment]
Hello, I am working on build this repo on Windows and AMD GPU... at https://github.com/sowson/darknet/tree/winX64-thanks-to-jihuacao on this branch... can you help to test it? I have to make sure macOS, GNU/Linux and Windows 10 x64 are build find before integrate, but maybe you have some experience and can improve this branch? Thanks!

**Comment by sowson** [comment]
@jihuacao I saw your fork... can you help? Thanks!

**Comment by sowson** [comment]
Pls, check https://iblog.isowa.io/2021/11/20/darknet-on-opencl-on-windows-11-x64 I only repeat steps by step guide and updated photo. It works now not sure why! ;-).
